src/errors.rs

```diff
@@ -0,0 +1,9 @@
+#[derive(Fail, Debug)]
```

Comment:
I strongly believe that we should implement `Error` instead of `Fail`, for a few reasons:

1. `failure` is ["experimental"](https://docs.rs/failure/0.1.5/failure/) and ["currently evolving"](https://github.com/rust-lang-nursery/failure#evolution), so while it may be fine for an end-user application, I don't think we should use it for a library like `ndarray-stats`.
2. The standard library's `Error` is being updated based on the `failure` experiment to resolve most of the issues with the old version of `Error`. ([RFC 2504](https://github.com/rust-lang/rfcs/blob/master/text/2504-fix-error.md), see esp. ["How this impacts failure"](https://github.com/rust-lang/rfcs/blob/master/text/2504-fix-error.md)) The [`.source()`](https://doc.rust-lang.org/std/error/trait.Error.html#method.source) method has been added, [`.cause()`](https://doc.rust-lang.org/std/error/trait.Error.html#method.cause) has been deprecated, and [`.description()`](https://doc.rust-lang.org/std/error/trait.Error.html#method.description) is "soft-deprecated".
3. It's annoying to have to call [`.compat()`](https://docs.rs/failure/0.1.5/failure/trait.Fail.html#method.compat) to get something that implements `Error`.

Implementing `Error` is pretty straightforward:

```rust
impl fmt::Display for ShapeMismatch {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        write!(f, "Array shapes do not match: {:?} and {:?}.", first_shape, second_shape)
    }
}

impl Error for ShapeMismatch {}
```

Reply:
I wasn't aware of the RFC, and it also looks [pretty close to landing](https://github.com/rust-lang/rust/issues/53487). Given the "How this impacts failure" section I agree it makes sense to implement `Error` - I'll fix it straight-away.